skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+                                   min(d + 1, n_z - z - 1)):
+                            z_start_k = z_start + k
+                            z_end_k = z_end + k
+                            weight = patch_distance_3d(
+                                    padded[x_start: x_end,
+                                           y_start: y_end,
+                                           z_start: z_end],
+                                    padded[x_start_i: x_end_i,
+                                           y_start_j: y_end_j,
+                                           z_start_k: z_end_k],
+                                    w, s)
+                            weight_sum += weight
+                            new_value += weight * padded[x + i, y + j, z + k]
+                result[x, y, z] = new_value / weight_sum
+    return result[offset:-offset, offset:-offset, offset:-offset]
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def _fast_nl_means_denoising_2d(image, int s=7, int d=13, float h=0.1):
+    """
+    Perform fast non-local means denoising on 2-D array, with the outer
+    loop on patch shifts in order to reduce the number of operations.
+
+    Parameters
+    ----------
+    image: ndarray
+        2-D input data to be denoised
+
+    s: int, optional
+        size of patches used for denoising
+
+    d: int, optional
+        maximal distance in pixels where to search patches used for denoising
+
+    h: float, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int offset = s / 2
+    # Image padding: we need to account for patch size, possible shift,
+    # + 1 for the boundary effects in finite differences
+    cdef int pad_size = offset + d + 1
+    cdef DTYPE_t [:, ::1] padded = np.ascontiguousarray(util.pad(image,
+                                pad_size, mode='reflect').astype(np.float32))
+    cdef DTYPE_t [:, ::1] result = np.zeros_like(padded)
+    cdef DTYPE_t [:, ::1] weights = np.zeros_like(padded)
+    cdef DTYPE_t [:, ::1] integral = np.zeros_like(padded)
+    cdef int n_x, n_y, t1, t2, x, y
+    cdef float weight, distance
+    cdef float alpha
+    cdef float h2 = h ** 2.
+    cdef float s2 = s ** 2.
+    n_x, n_y = image.shape
+    n_x += 2 * pad_size
+    n_y += 2 * pad_size
+    # Outer loops on patch shifts
+    # With t2 >= 0, reference patch is always on the left of test patch
+    for t1 in range(-d, d + 1):
+        for t2 in range(0, d + 1):
+            # alpha is to account for patches on the same column
+            # distance is computed twice in this case
+            if t2 == 0 and t1 is not 0:
+                alpha = 0.5
+            else:
+                alpha = 1.
+            integral = np.zeros_like(padded)
+            for x in range(max(1, - t1), min(n_x, n_x - t1)):
```

Comment:
A lot of the functions in this module are really huge... I think it would help a lot if some of these loops became their own inline functions. For example, here you could replace this block with `integral = _compute_patch_integral(padded, n_r, t1, n_c, t2)`